### PR TITLE
Adima

### PR DESCRIPTION
**Pull request by alindima**
Gata tema!!

### CODE REVIEW

**Comment by AndreiOprisan94** [comment]
Great! @alindima I will take a look as soon as I can.

File: DimaAlin/examples/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+out
```
Comment (AndreiOprisan94) [inline DimaAlin/examples/.gitignore]: There is no need to have a **.gitignore** file here. There is one in the root of the project that covers all of this patterns. Please remove this file.

File: DimaAlin/src/org/bootcamp/MainApp.java [hunk 2]
```diff
@@ -1,39 +1,22 @@
...
-
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileNotFoundException;
-import java.io.InputStream;
+
 import java.util.List;
-import java.util.Scanner;
 
 public class MainApp {
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/MainApp.java]: Please make this class **final**

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoDao.java [hunk 3]
```diff
@@ -6,4 +6,6 @@
 
 public interface VehicleInfoDao {
     List<VehicleInfo> getAllVehicles();
+
+    VehicleInfo getVehicleInfoById(String id);
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoDao.java]: Good 👍

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
+package org.bootcamp.dao;
+
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+import org.bootcamp.model.VehicleInfo;
+
+import java.io.*;
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: Please do not use wildcard imports. It is more clear if every symbol is declared explicitly.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.*;
+
+public final class VehicleInfoExcelFileDao implements VehicleInfoDao{
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: Please put a space between the name of the interface and the bracket

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    private static final int VEHICLE_IS_DIESEL = 5;
+    private static final int VEHICLE_ID = 0;
+    private static final int VEHICLE_FORMULA = 2;
+
+    private final String filePath;
+
+    private final Map<String, VehicleInfo> vehicleInfos;
+
+    public VehicleInfoExcelFileDao(String filePath) {
+        this.vehicleInfos = new LinkedHashMap<>();
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: Please initialize the **Map** at it's declaration. It is more clear that way.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        this.filePath = filePath;
+
+        try {
+            final InputStream inputStream = new FileInputStream(filePath);
+
+            Workbook workbook = new XSSFWorkbook(inputStream);
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: Please make every variable you do not mean to change **final**

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+            while (iterator.hasNext()) {
+
+                Row currentRow = iterator.next();
+
+                final VehicleInfo.Builder builder = VehicleInfo.builder();
+                final VehicleInfo info = builder.withId(currentRow.getCell(VEHICLE_ID).getStringCellValue())
+                        .withVehicleTypeName(currentRow.getCell(VEHICLE_TYPE).getStringCellValue())
+                        .withVehicleTypeFormula(currentRow.getCell(VEHICLE_FORMULA).getStringCellValue())
+                        .withAge((int)currentRow.getCell(VEHICLE_AGE).getNumericCellValue())
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: Please do not use C-style cast. Extract this in some variables and use `new Double(someDouble).int/longValue`.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                        .withNumberOfMiles((long)currentRow.getCell(VEHICLE_MILES).getNumericCellValue())
+                        .withDiesel(currentRow.getCell(VEHICLE_IS_DIESEL).getBooleanCellValue())
+                        .build();
+
+                vehicleInfos.put(info.getId(), info);
+            }
+        } catch (FileNotFoundException e) {
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: It is sufficient if you just catch the more generic **Exception** class. Only one `catch` statement. Please use `this.getClass().getName()` to obtain the name of the class.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            throw new IllegalStateException("Can not create instance of class VehicleInfoExcelFileDao");
+        } catch (IOException e) {
+            throw new IllegalStateException("Can not create instance of class VehicleInfoExcelFileDao");
+        }
+    }
+
+    @Override
+    public List<VehicleInfo> getAllVehicles() {
+        final List<VehicleInfo> vehicles = new ArrayList<>();
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoExcelFileDao.java]: You can do this in only one line. Please see the **values** method from Map and the constructor API for **ArrayList<>** class.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package org.bootcamp.dao;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.bootcamp.model.VehicleInfo;
+
+import java.io.*;
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java]: Same comment for wildcard imports.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.*;
+
+public final class VehicleInfoJsonFileDao implements VehicleInfoDao {
+
+    private final String filePath;
+
+    private final Map<String, VehicleInfo> vehicleInfos;
+
+    public VehicleInfoJsonFileDao(String filePath) {
+        this.vehicleInfos = new LinkedHashMap<>();
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java]: And for Map initialization

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            List<VehicleInfo> infos = objectMapper.readValue(inputStream, objectMapper.getTypeFactory()
+                    .constructCollectionType(List.class, VehicleInfo.class));
+
+            for(VehicleInfo info : infos){
+                vehicleInfos.put(info.getId(), info);
+            }
+
+
+        } catch (IOException e) {
+            throw new IllegalStateException("Can not create instance of class VehicleInfoJsonFileDao");
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java]: And for class name

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    @Override
+    public List<VehicleInfo> getAllVehicles() {
+        final List<VehicleInfo> vehicles = new ArrayList<>();
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoJsonFileDao.java]: Again. Please see how you can do this in just one line. I described above.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java [hunk 15]
```diff
@@ -6,10 +6,7 @@
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
 import java.io.InputStream;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Scanner;
+import java.util.*;
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java]: No wildcards plz.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java [hunk 16]
```diff
@@ -42,19 +37,38 @@ public VehicleInfoPlainFileDao(String filePath) {
...
 
-                vehicles.add(info);
+                vehicleInfos.put(tokens[VEHICLE_ID], info);
             }
 
             scanner.close();
 
         } catch (FileNotFoundException noFileFound) {
-            return Collections.emptyList();
+            throw new IllegalStateException("Can not create instance of class VehicleInfoPlainFileDao");
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java]: You know what I want here from the previous comments

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java [hunk 17]
```diff
@@ -42,19 +37,38 @@ public VehicleInfoPlainFileDao(String filePath) {
...
+        }
+    }
+
+    @Override
+    public List<VehicleInfo> getAllVehicles() {
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java]: And here.

File: DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java [hunk 18]
```diff
@@ -24,13 +21,11 @@
 
     private final String filePath;
 
+    private final Map<String, VehicleInfo> vehicleInfos;
+
     public VehicleInfoPlainFileDao(String filePath) {
+        this.vehicleInfos = new LinkedHashMap<>();
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/dao/VehicleInfoPlainFileDao.java]: Move the Map up please.

File: DimaAlin/src/org/bootcamp/formula/Formula.java [hunk 19]
```diff
@@ -38,7 +38,43 @@ public int calculate(Vehicle vehicle) {
 
             return cost;
         }
-    };
+    },
+
+    CAR_CHRISTMAS_FORMULA {
+        @Override
+        public int calculate(Vehicle vehicle) {
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/formula/Formula.java]: Please use the methods from the previous declared enum objects to get the initial cost and then apply the discount on the result. Please make the cost be **float** or **double** type. And modify in everyplace it is used. Discard ce C-style cast also.

File: DimaAlin/src/org/bootcamp/model/VehicleInfo.java [hunk 20]
```diff
@@ -11,7 +11,7 @@
     public VehicleInfo() {
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/model/VehicleInfo.java]: Please make this constructor **private** when using the **builder** pattern we do not want anybody create an instance of our class by directly using any constructor.

File: DimaAlin/src/org/bootcamp/service/InsuranceCalculatorService.java [hunk 21]
```diff
@@ -15,15 +17,16 @@
 public final class InsuranceCalculatorService {
     private String filePath;
```
Comment (AndreiOprisan94) [inline DimaAlin/src/org/bootcamp/service/InsuranceCalculatorService.java]: You do not need to keep the **filePath** as a class member anymore. Please remove it

**Comment by alindima** [comment]
I've made all the requested changes 😄

**Pull request closed without merging**